urlfcc351a1ba_kuntapavankumar_Z_tgz-pJ8-edu_concordia_dpis_ReplicaTestJ8/src/edu/concordia/dpis/messenger/UDPClient.java

```diff
@@ -0,0 +1,64 @@
+package edu.concordia.dpis.messenger;
+
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.net.SocketException;
+import java.net.SocketTimeoutException;
+import java.util.Arrays;
+
+import edu.concordia.dpis.commons.Imessenger;
+import edu.concordia.dpis.commons.Message;
+import edu.concordia.dpis.commons.MessageTransformer;
+import edu.concordia.dpis.commons.ReliableMessage;
+import edu.concordia.dpis.commons.TimeoutException;
+
+// UPD CLIENT to send messages over udp
+public class UDPClient implements Imessenger {
+
+	public static final UDPClient INSTANCE = new UDPClient();
+
+	@Override
+	public Message send(Message msg, int timeout)
+			throws edu.concordia.dpis.commons.TimeoutException {
+		String response = null;
+		DatagramSocket aSocket = null;
+		try {
+			aSocket = new DatagramSocket();
+			byte[] m = MessageTransformer.serializeMessage(msg);
+			InetAddress aHost = InetAddress.getByName(msg.getToAddress()
+					.getHost());
+			DatagramPacket request = new DatagramPacket(m, m.length, aHost, msg
+					.getToAddress().getPort());
+			aSocket.send(request);
+			if (msg.isReplyToThisMessage()) {
+				byte[] buffer = new byte[1000];
+				DatagramPacket reply = new DatagramPacket(buffer, buffer.length);
+				aSocket.setSoTimeout(timeout);
+				try {
+					aSocket.receive(reply);
+				} catch (SocketTimeoutException timeoutException) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)